src/crypto/rx-slow-hash.c

```diff
@@ -115,6 +115,46 @@ static inline int enabled_flags(void) {
 #define SEEDHASH_EPOCH_BLOCKS	2048	/* Must be same as BLOCKS_SYNCHRONIZING_MAX_COUNT in cryptonote_config.h */
 #define SEEDHASH_EPOCH_LAG		64
 
+static inline int is_power_of_2(uint64_t n) { return n && (n & (n-1)) == 0; }
+
+static uint64_t get_seedhash_epoch_lag()
+{
+  static uint64_t lag = (uint64_t)-1;
```

Comment:
The probability should be low, but this whole section is racy and would lead to unexpected values on platforms where a 64-bit write is not natively atomic. On those platforms, the 64-bit value is written in multiple CPU cycles, which can result in a read that is neither the initial nor final value ("tearing"). I'm not sure of the impact of getting a bogus value here (i.e. I don't know if the unlikely startup issue is recoverable or fatal), so take that into consideration with these suggestions.

The obvious fix is to make it (the C equivalent of) an `atomic<uint64_t>`. I wouldn't expect a performance decrease on any 64-bit platform we target (it only adds a memory fence that the compiler cannot optimize around), but I would expect the C compiler/runtime to add a lock around both the read and write on 32-bit platforms (except for x32 which we don't target).

The other fix is to use C11 [`call_once`](https://en.cppreference.com/w/c/thread/call_once). The downside is that all platforms will have a hidden `atomic<bool>` that gets checked before all reads to `lag` -> adding a memory fence + read + branch (that the compiler vendor will typically use `unlikely`, etc., to help with prediction). Support for this function might be spotty though.

A third technique use C++ to get the same effect as `call_once`. But you'd have to create a new C++ file with a `noexcept` function that has C linkage or something.

Reply:
This is trivially fixed by making the types int, and just adds a conversion at the use site, which is ~epsilon. int can be assumed atomic, so that would fix your objection ?

Reply:
No, a "naked" `int` cannot be assumed to be atomic either as per C/C++ rules. The compiler/runtime/hardware can do as it pleases for optimization purposes. We should "get away with it" on nearly all platform combinations that we target.

Cleaning this sort of thing up would be nice for better portability guarantees, but maybe that isn't practical right now if these tests are that important. An `atomic<int>` would cover all such pessimistic cases while adding minimal overhead (see my last comment) to the bulk of 32/64 bit platforms.

Reply:
Also, is `unsigned` more appropriate?

Reply:
It's C, no std::atomic. int's atomic pretty much everywhere. The GNU libc documentation says we can assume it in practice.

Reply:
C11 has an equivalent thing with a different interface.

Reply:
Also, we aren't targeting GNU libc 100% of the time, and may not be in the future either. Thus my comments about "getting away with it", yet not necessarily being ideal.

Reply:
BTW, I'm fine with using `int` (although again, `unsigned`?).

Reply:
Does it matter if it's libc or not, since it's an arch feature (unless you're implying GNU libc generates implicit locking code on some archs if int would not actually be atomic ?)
I'll make it unsigned.

Reply:
I'm not requesting any changes just clarifying for posterity on Github - it depends on the compiler/toolchain+architecture. I have no idea what GNU libc would do if the hardware cannot read/write a register-sized value in one cycle. Tell everyone the hardware is broken? The hardware would be antiquated at that point, but ISO C types will just say "I told you so!". Not losing sleep over this code though (now that its `unsigned` instead of `int64`).